src/lib/storage/dictionary_column.hpp

```diff
@@ -0,0 +1,192 @@
+  }
+}
+
+template <typename T>
+const AllTypeVariant DictionaryColumn<T>::operator[](const size_t i) const {
+  auto value_id = _attribute_vector->get(i);
+  return _dictionary->at(value_id);
+}
+
+template <typename T>
+const T DictionaryColumn<T>::get(const size_t i) const {
+  auto value_id = _attribute_vector->get(i);
+  return _dictionary->at(value_id);
+}
+
+template <typename T>
+void DictionaryColumn<T>::append(const AllTypeVariant&) {
+  throw std::logic_error("Dictionary columns are immutable!");
+}
+
+template <typename T>
+std::shared_ptr<const std::vector<T>> DictionaryColumn<T>::dictionary() const {
+  return _dictionary;
+}
+
+template <typename T>
+std::shared_ptr<const BaseAttributeVector> DictionaryColumn<T>::attribute_vector() const {
+  return _attribute_vector;
+}
+
+template <typename T>
+const T& DictionaryColumn<T>::value_by_value_id(ValueID value_id) const {
+  return _dictionary->at(value_id);
+}
+
+template <typename T>
+ValueID DictionaryColumn<T>::lower_bound(const T value) const {
+  DebugAssert(_dictionary, "Dictionary pointer is empty");
+  auto lower_bound_it = std::lower_bound(_dictionary->begin(), _dictionary->end(), value);
+  if (lower_bound_it == _dictionary->end()) return INVALID_VALUE_ID;
+  return ValueID(lower_bound_it - _dictionary->begin());
+}
+
+template <typename T>
+ValueID DictionaryColumn<T>::lower_bound(const AllTypeVariant& value) const {
+  auto casted_value = type_cast<T>(value);
+  DebugAssert(casted_value, "Value has wrong type");
+  return lower_bound(casted_value);
+}
+
+template <typename T>
+ValueID DictionaryColumn<T>::upper_bound(const T value) const {
+  DebugAssert(_dictionary, "Dictionary pointer is empty");
+  auto upper_bound_it = std::upper_bound(_dictionary->begin(), _dictionary->end(), value);
+  if (upper_bound_it == _dictionary->end()) return INVALID_VALUE_ID;
+  return ValueID(upper_bound_it - _dictionary->begin());
+}
+
+template <typename T>
+ValueID DictionaryColumn<T>::upper_bound(const AllTypeVariant& value) const {
+  auto casted_value = type_cast<T>(value);
+  DebugAssert(casted_value, "Value has wrong type");
+  return upper_bound(type_cast<T>(casted_value));
+}
+
+template <typename T>
+size_t DictionaryColumn<T>::unique_values_count() const {
+  return _dictionary->size();
+}
+
+template <typename T>
+size_t DictionaryColumn<T>::size() const {
+  return _attribute_vector->size();
+}
+
+template <typename T>
+std::shared_ptr<BaseAttributeVector> DictionaryColumn<T>::_get_fitted_vector(const size_t size) {
+  auto unique_count = dictionary()->size();
+
+  if (unique_count <= std::numeric_limits<uint8_t>::max()) {
+    return std::make_shared<FittedAttributeVector<uint8_t>>(size);
+  } else if (unique_count <= std::numeric_limits<uint16_t>::max()) {
+    return std::make_shared<FittedAttributeVector<uint16_t>>(size);
+  }
+
+  return std::make_shared<FittedAttributeVector<uint32_t>>(size);
```

Comment:
if size doesn't fit in uint32_t, you might get an overflow. Consider using an Assert